theano/gpuarray/__init__.py

```diff
@@ -101,6 +91,21 @@ def init_dev(dev, name=None):
                 ctx_props['cudnn_handle'] = dnn._make_handle(context)
             except Exception:
                 pass
+    if need_preallocate:
+        MB = (1024 * 1024)
+        if config.gpuarray.preallocate <= 1:
+            gmem = min(config.gpuarray.preallocate, 0.95) * ctx.total_gmem
+        else:
+            gmem = config.gpuarray.preallocate * MB
+        gmem = min(ctx.free_gmem - 50 * MB, gmem)
```

Comment:
On line 75, you should add an else clause. Otherwise, ctx won't always exist.

I know we discussed about this, but I don't think it is good to do that anymore without at least a warning. If user told that they want to preallocated 1000, they could end up with much less then that.

I would go to remove that new hard constraint and put a warning if there isn't 50MB available.

When we agree on the new algo, you also need to update the doc of the theano flag.

If people specify a